### PR TITLE
Switch over to console-feed for the Console component

### LINKED ISSUE #522: undefined is logged as null in the built-in console
`console.log(undefined)` results in `null` being printed in the built-in console, rather than `undefined` as expected.

### PR DESCRIPTION
**Pull request by samdenty99**
Closes #522 #408 #426 #374

image

**What kind of change does this PR introduce?**
Moving the Console handling over to the newly created [**`console-feed`**](http://npmjs.com/package/console-feed) package, to clean up the source code + add new features.

**What is the new behavior?**

The console has received an upgrade:
- DOM elements can now be logged
- Fixed inconsistent rendering
- `console.table` support
- Functions can be logged - names + function body
- Custom types can now be logged (eg. Promise)
- All objects can now be expanded to reveal a `<constructor>` item, showing the `constructor.name`

...and more stuff in the future. It's [super easy to setup transforms](https://github.com/samdenty99/console-feed/blob/master/src/module/Transform/HTML.ts)

**Checklist**:

- [ ] Documentation
- [ ] Tests
   - The package currently doesn't have any tests of it's own. If someone wants send a pr feel free
   - Possibly merge #703 
- [x] Ready to be merged
- [ ] Added myself to contributors table

### CODE REVIEW

File: packages/app/src/app/components/Preview/DevTools/Problems/index.js [hunk 1]
```diff
@@ -87,7 +89,10 @@ class Problems extends React.PureComponent<*, State> {
         {root && (
           <div>
             <File>Root</File>
-            {root.map((message, i) => (
+            {/*
+              What's going on here?
+              */
```
Comment (samdenty99) [inline packages/app/src/app/components/Preview/DevTools/Problems/index.js]: @CompuIves I couldn't get that component to display, so I wasn't able to integrate it. Under what circumstances is it displayed?

File: packages/app/package.json [hunk 2]
```diff
@@ -139,6 +140,7 @@
     "hash-sum": "^1.0.2",
     "humps": "CompuIves/humps",
     "immer": "^0.8.5",
+    "immutability-helper": "^2.6.6",
```
Comment (samdenty99) [inline packages/app/package.json]: (Possibly) not required. I've used it manually update the state, as calling `setState` multiple times in a row, to push to an array can

File: packages/app/src/app/components/Preview/DevTools/Console/Message/index.js [hunk 3]
```diff
@@ -1,112 +0,0 @@
...
-      if (msgArray.length > 1) {
-        otherErrorLines = msgArray.slice(1);
-      }
-
-      if (!otherErrorLines) {
-        return <InnerItem>{message.arguments.join(' ')}</InnerItem>;
-      }
-
-      return (
-        <InnerItem>
```
Comment (samdenty99) [inline packages/app/src/app/components/Preview/DevTools/Console/Message/index.js]: This needs to be re-implemented into `console-feed`

File: packages/app/src/app/components/Preview/DevTools/Console/index.js [hunk 4]
```diff
@@ -37,31 +37,50 @@ class Console extends React.Component {
...
       }
       case 'eval-result': {
         const { result, error } = data;
 
-        const parsedJson = result ? CircularJSON.parse(result) : result;
+        const decoded = Decode(result);
 
         if (!error) {
-          this.addMessage('log', [parsedJson], 'return');
+          this.addMessage('return', [decoded]);
```
Comment (samdenty99) [inline packages/app/src/app/components/Preview/DevTools/Console/index.js]: Currently the `return` & `command` types are unimplemented in `console-feed` so it defaults to a standard log

File: packages/app/src/app/components/Preview/DevTools/Console/index.js [hunk 5]
```diff
@@ -70,10 +89,10 @@ class Console extends React.Component {
               this.addMessage('warn', [t]);
             });
           } else {
-            this.addMessage('warn', [undefined], 'return');
+            this.addMessage('warn', [undefined]);
           }
         } else {
-          this.addMessage('log', [error]);
+          this.addMessage('error', [error]);
```
Comment (samdenty99) [inline packages/app/src/app/components/Preview/DevTools/Console/index.js]: Maybe you meant to do this? I thought it was a typo

**Comment by samdenty99** [comment]
2.4.0
- Command types
- Result types
- Custom object constructor types

image

File: packages/app/src/app/components/Preview/DevTools/Console/Message/index.js (hunk 3 above)
Comment (samdenty99) [reply packages/app/src/app/components/Preview/DevTools/Console/Message/index.js]: re-implemented :white_check_mark:

File: packages/app/src/app/components/Preview/DevTools/Console/index.js (hunk 4 above)
Comment (samdenty99) [reply packages/app/src/app/components/Preview/DevTools/Console/index.js]: re-implemented :white_check_mark:

File: packages/app/src/app/components/Preview/index.js [hunk 6]
```diff
@@ -176,7 +176,7 @@ class BasePreview extends React.Component<Props, State> {
     const sandbox = this.props.sandbox;
 
     if (settings.clearConsoleEnabled) {
-      console.clear(); // eslint-disable-line no-console
+      console.clear('__internal__'); // eslint-disable-line no-console
```
[the remark this answers is not in the document]
Comment (samdenty99) [reply packages/app/src/app/components/Preview/index.js]: image

I've added a message for every time the console is cleared. Adding the `__internal__` param to `console.clear()`  hides this message. The expected behaviour is that you only see `Console was cleared`, when you call `console.clear` or click the clear button

**Comment by samdenty99** [comment]
@CompuIves Ready to merge.

**Console tab:**

image

---

**Problems tab:**

image

---

**Different log type tests:**
<details><summary>Input commands</summary><p>

```js
console.log(`Console %c formatting`, 'color: red')
console.table('Table', [1, 2, 3])
console.log('Window', window)
console.log('Function with body', function a() { console.log(1) })
console.log('Mixed types', 1, 2)
console.log('inline', 'text')
console.log('Basic object', { an: 'object' })
console.log('Promise object', new Promise(() => { }))
console.log('Recursive object with promise', [[[['recursive', new Promise(() => { })]]]])

console.log('HTML element', document.body)
console.log('Named functions', [function myFunc() { }])

console.log('Nested', {
  promise: new Promise(() => { }),
  html: document.body
})

console.log('Falsey types', false, '', 0, null, undefined)
console.warn('Warning', 'message')
console.info('Info', 'message')
console.debug('Debug', 'message')

try {
  nonExistentFunc()
} catch (e) {
  console.error(e)
}
```

</p></details>

<br>

image

File: packages/app/src/sandbox/index.js [hunk 7]
```diff
@@ -54,6 +54,16 @@ requirePolyfills().then(() => {
...
+          // Attempt to wrap command in parentheses, fixing issues
+          // where directly returning objects results in unexpected
+          // behaviour.
+          try {
+            const wrapped = `(${data.command})`;
+            // `new Function` is used to validate Javascript syntax
+            const validate = new Function(wrapped);
+            data.command = wrapped;
+          } catch (e) {}
+
```
Comment (samdenty99) [inline packages/app/src/sandbox/index.js]: Before:

File: packages/app/src/app/components/Preview/DevTools/Problems/index.js (hunk 1 above)
Comment (CompuIves) [reply packages/app/src/app/components/Preview/DevTools/Problems/index.js]: Could we get this Root back? This is a reproducible case where it's needed: https://codesandbox.io/s/4061l2nqj0

Comment (samdenty99) [reply packages/app/src/app/components/Preview/DevTools/Problems/index.js]: @CompuIves Restored in 03c241d 


image

**Comment by samdenty99** [comment]
You might want to clean up the dependencies

```diff
- circular-json
- react-inspector

**Comment by CompuIves** [comment]
Great! Thank you so much for this @samdenty99, this fixes many issues. I'll clean those dependencies up on master and will do some testing on staging. Should be live this afternoon!

**Comment by samdenty99** [comment]
🙌

**Pull request merged**